# Designing and Creating a Baseball Database

## We will be working with four csv files describing Major League Baseball game statistics from Retrosheet:
* game_log.csv
* park_codes.csv
* person_codes.csv
* team_codes.csv

### Explanation of fields in game_log.csv:

In [1]:
# read in from command prompt
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


## Read in the data:

In [2]:
import pandas as pd
pd.set_option('max_columns',180)
pd.set_option('max_rows',200000)
pd.set_option('max_colwidth',5000)

game_log = pd.read_csv('game_log.csv')
park_codes = pd.read_csv('park_codes.csv')
person_codes = pd.read_csv('person_codes.csv')
team_codes = pd.read_csv('team_codes.csv')

/dataquest/system/env/python3/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


## Preliminary data exploration:

### game_log.csv

In [3]:
game_log.shape

(171907, 161)

In [4]:
game_log.head(15)

date  number_of_game day_of_week v_name v_league  v_game_number  \
0   18710504               0         Thu    CL1      NaN              1   
1   18710505               0         Fri    BS1      NaN              1   
2   18710506               0         Sat    CL1      NaN              2   
3   18710508               0         Mon    CL1      NaN              3   
4   18710509               0         Tue    BS1      NaN              2   
5   18710511               0         Thu    CH1      NaN              2   
6   18710513               0         Sat    WS3      NaN              2   
7   18710513               0         Sat    CH1      NaN              3   
8   18710515               0         Mon    WS3      NaN              3   
9   18710516               0         Tue    TRO      NaN              2   
10  18710516               0         Tue    WS3      NaN              4   
11  18710517               0         Wed    WS3      NaN              5   
12  18710518               0         Thu    NY2      NaN              1   
13  18710519               0         Fri    WS3      NaN              6   
14  18710520               0         Sat    PH1      NaN              1   

   h_name h_league  h_game_number  v_score  h_score  length_outs day_night  \
0     FW1      NaN              1        0        2         54.0         D   
1     WS3      NaN              1       20       18         54.0         D   
2     RC1      NaN              1       12        4         54.0         D   
3     CH1      NaN              1       12       14         54.0         D   
4     TRO      NaN              1        9        5         54.0         D   
5     CL1      NaN              4       18       10         48.0         D   
6     CL1      NaN              5       12        8         54.0         D   
7     FW1      NaN              2       14        5         54.0         D   
8     FW1      NaN              3        6       12         54.0         D   
9     BS1      NaN              3       29       14         54.0         D   
10    CH1      NaN              4        4       14         54.0         D   
11    RC1      NaN              2       12       15         54.0         D   
12    TRO      NaN              3       14        3         54.0         D   
13    CH1      NaN              5        7        9         54.0         D   
14    BS1      NaN              4        8       11         54.0         D   

   completion forefeit protest park_id  attendance  length_minutes  \
0         NaN      NaN     NaN   FOR01       200.0           120.0   
1         NaN      NaN     NaN   WAS01      5000.0           145.0   
2         NaN      NaN     NaN   RCK01      1000.0           140.0   
3         NaN      NaN     NaN   CHI01      5000.0           150.0   
4         NaN      NaN     NaN   TRO01      3250.0           145.0   
5         NaN        V     NaN   CLE01      2500.0           120.0   
6         NaN      NaN     NaN   CIN01      1200.0           150.0   
7         NaN      NaN     NaN   FOR01      1500.0           105.0   
8         NaN      NaN     NaN   FOR01         NaN           140.0   
9         NaN      NaN     NaN   BOS01      2500.0             NaN   
10        NaN      NaN     NaN   CHI01      5000.0           117.0   
11        NaN        V     NaN   RCK01         NaN           135.0   
12        NaN      NaN     NaN   TRO01      2000.0           120.0   
13        NaN      NaN     NaN   CHI01      5000.0           110.0   
14        NaN      NaN     NaN   BOS01      2500.0           155.0   

    v_line_score h_line_score  v_at_bats  v_hits  v_doubles  v_triples  \
0      000000000    010010000       30.0     4.0        1.0        0.0   
1      107000435    640113030       41.0    13.0        1.0        2.0   
2      610020003    010020100       49.0    11.0        1.0        1.0   
3      101403111    077000000       46.0    15.0        2.0        1.0   
4      000002232    101003000       46.0    17.0        4.0        

In [5]:
new_col = ( game_log["date"].astype(str) 
           + game_log["v_name"]
           + game_log["h_name"]
           + game_log["h_game_number"].astype(str)
          )
len(pd.unique(new_col))

171907

In [6]:
len(game_log["park_id"].dropna().unique())

245

In [7]:
game_log.describe(include="all")

date  number_of_game day_of_week  v_name v_league  \
count   1.719070e+05   171907.000000      171907  171907   170821   
unique           NaN             NaN           7     147        6   
top              NaN             NaN         Sat     CHN       NL   
freq             NaN             NaN       28891    8870    88866   
mean    1.953462e+07        0.260897         NaN     NaN      NaN   
std     4.149326e+05        0.605667         NaN     NaN      NaN   
min     1.871050e+07        0.000000         NaN     NaN      NaN   
25%     1.918052e+07        0.000000         NaN     NaN      NaN   
50%     1.953053e+07        0.000000         NaN     NaN      NaN   
75%     1.989051e+07        0.000000         NaN     NaN      NaN   
max     2.016100e+07        3.000000         NaN     NaN      NaN   

        v_game_number  h_name h_league  h_game_number        v_score  \
count   171907.000000  171907   170821  171907.000000  171907.000000   
unique            NaN     147        6            NaN            NaN   
top               NaN     CHN       NL            NaN            NaN   
freq              NaN    9024    88867            NaN            NaN   
mean        76.929887     NaN      NaN      76.953806       4.420582   
std         45.178029     NaN      NaN      45.162564       3.278489   
min          1.000000     NaN      NaN       1.000000       0.000000   
25%         38.000000     NaN      NaN      38.000000       2.000000   
50%         76.000000     NaN      NaN      76.000000       4.000000   
75%        115.000000     NaN      NaN     115.000000       6.000000   
max        165.000000     NaN      NaN     165.000000      49.000000   

              h_score    length_outs day_night        completion forefeit  \
count   171907.000000  140841.000000    140150               116      145   
unique            NaN            NaN         2               116        3   
top               NaN            NaN         D  20150718,,2,3,30        H   
freq              NaN            NaN     82724                 1       69   
mean         4.701461      53.619976       NaN               NaN      NaN   
std          3.355605       5.571512       NaN               NaN      NaN   
min          0.000000       0.000000       NaN               NaN      NaN   
25%          2.000000      51.000000       NaN               NaN      NaN   
50%          4.000000      54.000000       NaN               NaN      NaN   
75%          6.000000      54.000000       NaN               NaN      NaN   
max         38.000000     156.000000       NaN               NaN      NaN   

       protest park_id     attendance  length_minutes v_line_score  \
count      180  171907  118877.000000   136701.000000       147271   
unique       5     245            NaN             NaN        36367   
top          V   STL07            NaN             NaN    000000000   
freq        90    7022            NaN             NaN        10102   
mean       NaN     NaN   20184.247188      150.903329          NaN   
std        NaN     NaN   14257.381902       34.748160          NaN   
min        NaN     NaN       0.000000       24.000000          NaN   
25%        NaN     NaN    7962.000000      125.000000          NaN   
50%        NaN     NaN   18639.000000      150.000000          NaN   
75%        NaN     NaN   31242.000000      173.000000          NaN   
max        NaN     NaN   99027.000000     1245.000000          NaN   

       h_line_score      v_at_bats         v_hits      v_doubles  \
count        147271  140838.000000  140838.000000  140772.000000   
unique        37859            NaN            NaN            NaN   
top       000000000            NaN            NaN            NaN   
freq           8028            NaN            NaN            NaN   
mean            NaN      34.914398       8.999318       1.563592   
std             NaN       4.633761       3.599728       1.346740   
min             NaN      12.000000       0.000000       0.000000   
25%         

### Game_log.csv contains game details and other statistics (offensive/defensive/pitching) on baseball games from 1871 to 2016. It is unique by game date, visiting team, home team, and home team game number (but there is no natural primary key). There are missing values and some non-sensical values (e.g. negative values for v_triples, v_homeruns, etc.).

### park_codes.csv

In [8]:
park_codes.shape

(252, 9)

In [9]:
park_codes.head()

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

In [10]:
len(park_codes["park_id"].unique())

252

In [11]:
park_codes.describe(include="all")

park_id           name             aka          city state       start  \
count      252            252              58           252   252         252   
unique     252            241              56            85    36         215   
top      TOL01  Athletic Park  Beyerle's Park  Philadelphia    NY  05/01/1883   
freq         1              4               2            14    40           5   

               end league                        notes  
count          222    186                          128  
unique         205      6                          123  
top     10/03/1915     NL  CL3:7/22&7/29&8/26&9/2/1888  
freq             3     88                            2

### Park_codes.csv contains infromation like the location and operation dates for 252 baseball parks. This dataset can be related to the main game_log.csv dataset using the park_id variable.

### person_codes.csv

In [12]:
person_codes.shape

(20494, 7)

In [13]:
person_codes.head()

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

In [14]:
person_codes.describe(include="all")

id   last  first player_debut   mgr_debut coach_debut  \
count      20494  20494  20433        19183         701        1566   
unique     20494  10397   2484        10278         537         332   
top     brazc001  Smith   Bill   05/01/1884  05/01/1884  04/01/2013   
freq           1    168    579           36          10          29   

         ump_debut  
count         1546  
unique        1130  
top     08/25/1978  
freq            52

### Person_codes.csv includes information on each player/coach/umpire debut date. The main dataset (game_log.csv) contains player, pitcher, and umpire names and IDs for the home and visiting teams, which can be merged with the id, last, and first variables from person_codes.csv.

### team_codes.csv

In [15]:
team_codes.shape

(150, 8)

In [16]:
team_codes.head()

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

In [17]:
sorted(team_codes["team_id"].unique())

['ALT',
 'ANA',
 'ARI',
 'ATL',
 'BAL',
 'BFN',
 'BFP',
 'BL1',
 'BL2',
 'BL4',
 'BLA',
 'BLF',
 'BLN',
 'BLU',
 'BOS',
 'BR1',
 'BR2',
 'BR3',
 'BR4',
 'BRF',
 'BRO',
 'BRP',
 'BS1',
 'BS2',
 'BSN',
 'BSP',
 'BSU',
 'BUF',
 'CAL',
 'CH1',
 'CH2',
 'CHA',
 'CHF',
 'CHN',
 'CHP',
 'CHU',
 'CIN',
 'CL1',
 'CL2',
 'CL3',
 'CL4',
 'CL5',
 'CL6',
 'CLE',
 'CLP',
 'CN1',
 'CN2',
 'CN3',
 'CNU',
 'COL',
 'DET',
 'DTN',
 'ELI',
 'FLO',
 'FW1',
 'HAR',
 'HOU',
 'HR1',
 'IN1',
 'IN2',
 'IN3',
 'IND',
 'KC1',
 'KC2',
 'KCA',
 'KCF',
 'KCN',
 'KCU',
 'KEO',
 'LAA',
 'LAN',
 'LS1',
 'LS2',
 'LS3',
 'MIA',
 'MID',
 'MIL',
 'MIN',
 'ML2',
 'ML3',
 'MLA',
 'MLN',
 'MLU',
 'MON',
 'NEW',
 'NH1',
 'NY1',
 'NY2',
 'NY3',
 'NY4',
 'NYA',
 'NYN',
 'NYP',
 'OAK',
 'PH1',
 'PH2',
 'PH3',
 'PH4',
 'PHA',
 'PHI',
 'PHN',
 'PHP',
 'PHU',
 'PIT',
 'PRO',
 'PT1',
 'PTF',
 'PTP',
 'PTU',
 'RC1',
 'RC2',
 'RIC',
 'SDN',
 'SE1',
 'SEA',
 'SFN',
 'SL1',
 'SL2',
 'SL3',
 'SL4',
 'SL5',
 'SLA',
 'SLF',
 'SLN',
 'SLU',


In [18]:
sorted(game_log["v_name"].unique())

['ALT',
 'ANA',
 'ARI',
 'ATL',
 'BAL',
 'BFN',
 'BFP',
 'BL1',
 'BL2',
 'BL4',
 'BLA',
 'BLF',
 'BLN',
 'BLU',
 'BOS',
 'BR1',
 'BR2',
 'BR3',
 'BR4',
 'BRF',
 'BRO',
 'BRP',
 'BS1',
 'BS2',
 'BSN',
 'BSP',
 'BSU',
 'BUF',
 'CAL',
 'CH1',
 'CH2',
 'CHA',
 'CHF',
 'CHN',
 'CHP',
 'CHU',
 'CIN',
 'CL1',
 'CL2',
 'CL3',
 'CL4',
 'CL5',
 'CL6',
 'CLE',
 'CLP',
 'CN1',
 'CN2',
 'CN3',
 'CNU',
 'COL',
 'DET',
 'DTN',
 'ELI',
 'FLO',
 'FW1',
 'HAR',
 'HOU',
 'HR1',
 'IN1',
 'IN2',
 'IN3',
 'IND',
 'KC1',
 'KC2',
 'KCA',
 'KCF',
 'KCN',
 'KCU',
 'KEO',
 'LAA',
 'LAN',
 'LS1',
 'LS2',
 'LS3',
 'MIA',
 'MID',
 'MIL',
 'MIN',
 'ML2',
 'ML3',
 'MLA',
 'MLN',
 'MON',
 'NEW',
 'NH1',
 'NY1',
 'NY2',
 'NY3',
 'NY4',
 'NYA',
 'NYN',
 'NYP',
 'OAK',
 'PH1',
 'PH2',
 'PH3',
 'PH4',
 'PHA',
 'PHI',
 'PHN',
 'PHP',
 'PHU',
 'PIT',
 'PRO',
 'PT1',
 'PTF',
 'PTP',
 'RC1',
 'RC2',
 'RIC',
 'SDN',
 'SE1',
 'SEA',
 'SFN',
 'SL1',
 'SL2',
 'SL3',
 'SL4',
 'SL5',
 'SLA',
 'SLF',
 'SLN',
 'SLU',
 'SPU',
 'SR1',


In [19]:
team_codes.describe(include="all")

team_id league        start          end        city nickname  \
count      150    124   150.000000   150.000000         150      150   
unique     149      6          NaN          NaN          54       93   
top        MIL     NL          NaN          NaN  Washington     Reds   
freq         2     45          NaN          NaN          12        6   
mean       NaN    NaN  1902.733333  1517.440000         NaN      NaN   
std        NaN    NaN    37.326002   761.772866         NaN      NaN   
min        NaN    NaN  1871.000000     0.000000         NaN      NaN   
25%        NaN    NaN  1879.000000  1872.000000         NaN      NaN   
50%        NaN    NaN  1887.000000  1884.000000         NaN      NaN   
75%        NaN    NaN  1911.250000  1891.000000         NaN      NaN   
max        NaN    NaN  2012.000000  2011.000000         NaN      NaN   

       franch_id         seq  
count        150  150.000000  
unique       115         NaN  
top          BS1         NaN  
freq           4         NaN  
mean         NaN    1.293333  
std          NaN    0.585584  
min          NaN    1.000000  
25%          NaN    1.000000  
50%          NaN    1.000000  
75%          NaN    1.000000  
max          NaN    4.000000

In [20]:
team_codes[team_codes.duplicated(subset="team_id", keep=False) == True]

team_id league  start   end       city nickname franch_id  seq
112     MIL     AL   1970  1997  Milwaukee  Brewers       SE1    2
113     MIL     NL   1998     0  Milwaukee  Brewers       SE1    3

### Team_codes.csv contains nicknames, duration, the location, and the league for different baseball teams. (Note that one baseball team_id=MIL is repeated twice in this dataset due to a change in its league.) This dataset can be merged to the variables v_name and h_name in the main game_log.csv dataset, using the team_id variable in team_codes.csv.  

## Start creating the SQL database

In [21]:
import sqlite3

db = 'mlb.db'

def run_query(query):
    '''takes in a SQL query and returns a pandas dataframe of the query'''
    with sqlite3.connect(db) as conn:
        return pd.read_sql(query,conn)
    
def run_command(command):
    '''takes in a SQL command and executes it'''
    with sqlite3.connect(db) as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(command)
        
def show_tables():
    '''calls run_query to return a list of all tables and views in the database'''
    query = '''SELECT name, type 
               FROM sqlite_master 
               WHERE type IN ('table','view');'''
    return run_query(query)

In [22]:
with sqlite3.connect(db) as conn:
    game_log.to_sql('game_log',conn,index=False,if_exists='replace')
    park_codes.to_sql('park_codes',conn,index=False,if_exists='replace')
    person_codes.to_sql('person_codes',conn,index=False,if_exists='replace')
    team_codes.to_sql('team_codes',conn,index=False,if_exists='replace')

In [23]:
show_tables()

name   type
0              person  table
1                park  table
2              league  table
3                game  table
4                team  table
5     team_appearance  table
6     appearance_type  table
7   person_appearance  table
8            game_log  table
9          park_codes  table
10       person_codes  table
11         team_codes  table

In [24]:
# add new variable to game_log to be used as the primary key

c = '''
    ALTER TABLE game_log
    ADD COLUMN game_id TEXT;
    '''

run_command(c)

c = '''
    UPDATE game_log
    SET game_id = h_name || date || number_of_game;
    '''

run_command(c)

In [25]:
# check that the variable creation looks okay

query = '''
    SELECT game_id, h_name, date, number_of_game
    FROM game_log
    LIMIT 5;
    '''
run_query(query)

game_id h_name      date  number_of_game
0  FW1187105040    FW1  18710504               0
1  WS3187105050    WS3  18710505               0
2  RC1187105060    RC1  18710506               0
3  CH1187105080    CH1  18710508               0
4  TRO187105090    TRO  18710509               0

## Looking for opportunities to Normalize the data

* player IDs, names, and def_pos's, as well as pitchers, managers, and umpires in the game_log dataset are attributes of the people, not the games. This information might be better placed in the person_codes dataset or in a separate 'game_players' dataset.

* The start and end columns in the park_codes dataset could be derived by looking at the date data in the game_log dataset, so it is redundant.

* Debut dates in the person_codes dataset could be derived by looking at the date data in the game_log dataset, so this is also redundant.

## Desiging a schema for the baseball database using [DBDesigner.net](https://www.dbdesigner.net/)

### first version:

![baseball_schema_v1](C:\Users\C\Desktop\datascience.io\baseball_schema_v1.jpg)

### final version:

![baseball_schema_v2](C:\Users\C\Desktop\datascience.io\baseball_schema_v2.jpg)

In [26]:
# create person table

c1 = '''
    CREATE TABLE IF NOT EXISTS person (
    person_id TEXT PRIMARY KEY,
    first_name TEXT,
    last_name TEXT)'''

c2 = '''
    INSERT OR IGNORE INTO person
    SELECT  id,
            first,
            last
    FROM person_codes'''

run_command(c1)
run_command(c2)

q = '''
    SELECT *
    FROM person
    LIMIT 5'''

run_query(q)

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [27]:
# create park table

c1 = '''
    CREATE TABLE IF NOT EXISTS park (
    park_id TEXT PRIMARY KEY,
    name TEXT,
    nickname TEXT,
    city TEXT,
    state TEXT,
    notes TEXT)'''

c2 = '''
    INSERT OR IGNORE INTO park
    SELECT  park_id,
            name,
            aka,
            city,
            state,
            notes
    FROM park_codes'''

run_command(c1)
run_command(c2)

q = '''
    SELECT *
    FROM park
    LIMIT 5'''

run_query(q)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [28]:
# create league table

c1 = '''
    CREATE TABLE IF NOT EXISTS league (
    league_id TEXT PRIMARY KEY,
    name TEXT)'''

c2 = '''
    INSERT OR IGNORE INTO league
    VALUES
        ("NL", "National League"),
        ("AL", "American League"),
        ("AA", "American Association"),
        ("FL", "Federal League"),
        ("PL", "Players League"),
        ("UA", "Union Association")'''

run_command(c1)
run_command(c2)

q = '''
    SELECT *
    FROM league
    LIMIT 5'''

run_query(q)

league_id                  name
0        NL       National League
1        AL       American League
2        AA  American Association
3        FL        Federal League
4        PL        Players League

In [39]:
# create appearance type table

c1 = '''DROP TABLE IF EXISTS appearance_type'''

run_command(c1)

c2 = '''
CREATE TABLE appearance_type (
    appearance_type_id TEXT PRIMARY KEY,
    name TEXT,
    category TEXT
)
'''
run_command(c2)

appearance_type_csv = pd.read_csv('appearance_type.csv')

with sqlite3.connect('mlb.db') as conn:
    appearance_type_csv.to_sql('appearance_type',
                           conn,
                           index=False,
                           if_exists='append')

In [40]:
q = '''
    SELECT *
    FROM appearance_type
    '''

run_query(q)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [41]:
# create team table

c1 = '''
    CREATE TABLE IF NOT EXISTS team (
    team_id TEXT PRIMARY KEY,
    league_id TEXT,
    city TEXT,
    nickname TEXT,
    franch_id TEXT,
    FOREIGN KEY (league_id) REFERENCES league(league_id))'''

c2 = '''
    INSERT OR IGNORE INTO team
    SELECT  team_id,
            league as league_id,
            city,
            nickname,
            franch_id
    FROM team_codes'''

run_command(c1)
run_command(c2)

q = '''
    SELECT *
    FROM team
    LIMIT 5'''

run_query(q)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [42]:
# create game table

c1 = '''
    CREATE TABLE IF NOT EXISTS game (
    game_id TEXT PRIMARY KEY,
    date TEXT,
    number_of_game INTEGER,
    park_id TEXT,
    length_outs INTEGER,
    day BOOLEAN,
    completion TEXT,
    forfeit TEXT,
    protest TEXT,
    attendance INTEGER,
    length_minutes INTEGER,
    additional_info TEXT,
    acquisition_info TEXT,
    FOREIGN KEY (park_id) REFERENCES park(park_id)
    )'''

c2 = '''
    INSERT OR IGNORE INTO game
    SELECT 
        game_id,
        date,
        number_of_game,
        park_id,
        length_outs,
        CASE 
            WHEN day_night = "D" THEN 1
            WHEN day_night = "N" THEN 0
            ELSE NULL
            END 
            as day,
        completion,
        forefeit,
        protest,
        attendance,
        length_minutes,
        additional_info,
        acquisition_info
    FROM game_log'''

run_command(c1)
run_command(c2)

q = '''
    SELECT *
    FROM game
    LIMIT 5'''

run_query(q)

game_id      date  number_of_game park_id  length_outs  day  \
0  FW1187105040  18710504               0   FOR01           54    1   
1  WS3187105050  18710505               0   WAS01           54    1   
2  RC1187105060  18710506               0   RCK01           54    1   
3  CH1187105080  18710508               0   CHI01           54    1   
4  TRO187105090  18710509               0   TRO01           54    1   

  completion forfeit protest  attendance  length_minutes additional_info  \
0       None    None    None         200             120            None   
1       None    None    None        5000             145            HTBF   
2       None    None    None        1000             140            None   
3       None    None    None        5000             150            None   
4       None    None    None        3250             145            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y

In [43]:
# create team_appearance table

c1 = '''
CREATE TABLE IF NOT EXISTS team_appearance (
    team_id TEXT,
    game_id TEXT,
    home BOOLEAN,
    league_id TEXT,
    score INTEGER,
    line_score TEXT,
    at_bats INTEGER,
    hits INTEGER,
    doubles INTEGER,
    triples INTEGER,
    homeruns INTEGER,
    rbi INTEGER,
    sacrifice_hits INTEGER,
    sacrifice_flies INTEGER,
    hit_by_pitch INTEGER,
    walks INTEGER,
    intentional_walks INTEGER,
    strikeouts INTEGER,
    stolen_bases INTEGER,
    caught_stealing INTEGER,
    grounded_into_double INTEGER,
    first_catcher_interference INTEGER,
    left_on_base INTEGER,
    pitchers_used INTEGER,
    individual_earned_runs INTEGER,
    team_earned_runs INTEGER,
    wild_pitches INTEGER,
    balks INTEGER,
    putouts INTEGER,
    assists INTEGER,
    errors INTEGER,
    passed_balls INTEGER,
    double_plays INTEGER,
    triple_plays INTEGER,
    PRIMARY KEY (team_id, game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (league_id) REFERENCES league(league_id)
)
'''

c2 = '''
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league,
        h_score,
        h_line_score,
        h_at_bats,
        h_hits,
        h_doubles,
        h_triples,
        h_homeruns,
        h_rbi,
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks,
        h_intentional_walks,
        h_strikeouts,
        h_stolen_bases,
        h_caught_stealing,
        h_grounded_into_double,
        h_first_catcher_interference,
        h_left_on_base,
        h_pitchers_used,
        h_individual_earned_runs,
        h_team_earned_runs,
        h_wild_pitches,
        h_balks,
        h_putouts,
        h_assists,
        h_errors,
        h_passed_balls,
        h_double_plays,
        h_triple_plays
    FROM game_log

UNION

    SELECT    
        v_name,
        game_id,
        0 AS home,
        v_league,
        v_score,
        v_line_score,
        v_at_bats,
        v_hits,
        v_doubles,
        v_triples,
        v_homeruns,
        v_rbi,
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks,
        v_intentional_walks,
        v_strikeouts,
        v_stolen_bases,
        v_caught_stealing,
        v_grounded_into_double,
        v_first_catcher_interference,
        v_left_on_base,
        v_pitchers_used,
        v_individual_earned_runs,
        v_team_earned_runs,
        v_wild_pitches,
        v_balks,
        v_putouts,
        v_assists,
        v_errors,
        v_passed_balls,
        v_double_plays,
        v_triple_plays
    FROM game_log
'''

run_command(c1)
run_command(c2)

q = '''
    SELECT * FROM team_appearance
    WHERE game_id = "FW1187105040"
    ORDER By game_id, home
    '''

run_query(q)

team_id       game_id  home league_id  score line_score  at_bats  hits  \
0     CL1  FW1187105040     0      None      0  000000000       30     4   
1     FW1  FW1187105040     1      None      2  010010000       31     4   

   doubles  triples  homeruns  rbi  sacrifice_hits  sacrifice_flies  \
0        1        0         0    0               0                0   
1        1        0         0    2               0                0   

   hit_by_pitch  walks intentional_walks  strikeouts  stolen_bases  \
0             0      1              None           6             1   
1             0      1              None           0             0   

  caught_stealing  grounded_into_double first_catcher_interference  \
0            None                    -1                       None   
1            None                    -1                       None   

   left_on_base  pitchers_used  individual_earned_runs  team_earned_runs  \
0             4              1                       1                 1   
1             3              1                       0                 0   

   wild_pitches  balks  putouts  assists  errors  passed_balls  double_plays  \
0             0      0       27        9       0             3             0   
1             0      0       27        3       3             1             1   

   triple_plays  
0             0  
1             0

In [44]:
# create person appearance table

c = '''
    DROP TABLE IF EXISTS person_appearance'''
run_command(c)

c1 = '''
    CREATE TABLE IF NOT EXISTS person_appearance (
    appearance_id INTEGER PRIMARY KEY,
    person_id TEXT,
    team_id TEXT,
    game_id TEXT,
    appearance_type_id,
    FOREIGN KEY (person_id) REFERENCES person(person_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id)
    )'''

# umpire, managers, pitchers, rbi

c2 = '''
    INSERT INTO person_appearance (
        game_id,
        team_id,
        person_id,
        appearance_type_id
    )

        SELECT
            game_id,
            NULL,
            hp_umpire_id,
            "UHP"
        FROM game_log
        WHERE hp_umpire_id IS NOT NULL    

    UNION

        SELECT
            game_id,
            NULL,
            [1b_umpire_id],
            "U1B"
        FROM game_log
        WHERE "1b_umpire_id" IS NOT NULL

    UNION

        SELECT
            game_id,
            NULL,
            [2b_umpire_id],
            "U2B"
        FROM game_log
        WHERE [2b_umpire_id] IS NOT NULL

    UNION

        SELECT
            game_id,
            NULL,
            [3b_umpire_id],
            "U3B"
        FROM game_log
        WHERE [3b_umpire_id] IS NOT NULL

    UNION

        SELECT
            game_id,
            NULL,
            lf_umpire_id,
            "ULF"
        FROM game_log
        WHERE lf_umpire_id IS NOT NULL

    UNION

        SELECT
            game_id,
            NULL,
            rf_umpire_id,
            "URF"
        FROM game_log
        WHERE rf_umpire_id IS NOT NULL

    UNION

        SELECT
            game_id,
            v_name,
            v_manager_id,
            "MM"
        FROM game_log
        WHERE v_manager_id IS NOT NULL

    UNION

        SELECT
            game_id,
            h_name,
            h_manager_id,
            "MM"
        FROM game_log
        WHERE h_manager_id IS NOT NULL

    UNION

        SELECT
            game_id,
            CASE
                WHEN h_score > v_score THEN h_name
                ELSE v_name
                END,
            winning_pitcher_id,
            "AWP"
        FROM game_log
        WHERE winning_pitcher_id IS NOT NULL

    UNION

        SELECT
            game_id,
            CASE
                WHEN h_score < v_score THEN h_name
                ELSE v_name
                END,
            losing_pitcher_id,
            "ALP"
        FROM game_log
        WHERE losing_pitcher_id IS NOT NULL

    UNION

        SELECT
            game_id,
            CASE
                WHEN h_score > v_score THEN h_name
                ELSE v_name
                END,
            saving_pitcher_id,
            "ASP"
        FROM game_log
        WHERE saving_pitcher_id IS NOT NULL

    UNION

        SELECT
            game_id,
            CASE
                WHEN h_score > v_score THEN h_name
                ELSE v_name
                END,
            winning_rbi_batter_id,
            "AWB"
        FROM game_log
        WHERE winning_rbi_batter_id IS NOT NULL

    UNION

        SELECT
            game_id,
            v_name,
            v_starting_pitcher_id,
            "PSP"
        FROM game_log
        WHERE v_starting_pitcher_id IS NOT NULL

    UNION

        SELECT
            game_id,
            h_name,
            h_starting_pitcher_id,
            "PSP"
        FROM game_log
        WHERE h_starting_pitcher_id IS NOT NULL
    '''

run_command(c1)
run_command(c2)

# offense and defense
template = '''
INSERT INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
'''

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
        # run commmand is a helper function which runs
        # a query against our database.
        run_command(template.format(**query_vars))

In [45]:
print(run_query("SELECT COUNT(DISTINCT game_id) as games_game FROM game"))
print(run_query("SELECT COUNT(DISTINCT game_id) as games_person_appearance FROM person_appearance"))

q = '''
SELECT
    pa.*,
    at.name,
    at.category
FROM person_appearance pa
INNER JOIN appearance_type at on at.appearance_type_id = pa.appearance_type_id
WHERE PA.game_id = (
                   SELECT max(game_id)
                    FROM person_appearance
                   )
ORDER BY team_id, appearance_type_id
'''

run_query(q)

   games_game
0      171907
   games_person_appearance
0                   171907


appearance_id person_id team_id       game_id appearance_type_id  \
0        1646114  steab101    None  WSU188409250                UHP   
1        1646116  murnt101     BSU  WSU188409250                 MM   
2        1646115  crane101     BSU  WSU188409250                PSP   
3        1646118  scanm801     WSU  WSU188409250                 MM   
4        1646117  dailh101     WSU  WSU188409250                PSP   

               name category  
0        Home Plate   umpire  
1           Manager  manager  
2  Starting Pitcher  pitcher  
3           Manager  manager  
4  Starting Pitcher  pitcher

In [46]:
# delete original datasets which are no longer needed

c1 = '''DROP TABLE IF EXISTS game_log'''
c2 = '''DROP TABLE IF EXISTS park_codes'''
c3 = '''DROP TABLE IF EXISTS team_codes'''
c4 = '''DROP TABLE IF EXISTS person_codes'''

run_command(c1)
run_command(c2)
run_command(c3)
run_command(c4)

show_tables()

name   type
0             person  table
1               park  table
2             league  table
3               game  table
4               team  table
5    team_appearance  table
6    appearance_type  table
7  person_appearance  table